Add scripts for using `git bisect`

xylar opened the pull request:
A config file determines the "good" and "bad" commits, and provides the paths to the MPAS code, the load script, etc.  It
also provides the commands for building the code, setting up the test case(s) and running them.

A driver script sets things up and makes the calls to `git bisect`, passing along the `utils/bisect/bisect_step.py` script for running each step of the bisection.

The `utils/bisect/bisect_step.py` script updates and builds the MPAS code, then sets up the compass test case(s) and runs compass.

xylar commented:
@sbrus89, I believe this is working now.  It is ready for you to test and review.  I don't plan to document it beyond `bisect/README.md` unless you would like it to be in the main documentation.

sbrus89 commented:
@xylar, very cool! I'll take a look.

xylar commented:
I've been trying to clean this up a bit (e.g. make sure the `make` output doesn't overwhelm everything else).  I'll re-test tomorrow or Monday.

xylar commented:
Using this approach, I was able to trace the unexpected non-BFB changes in baroclinic channel runs in https://github.com/MPAS-Dev/compass/pull/326 to https://github.com/E3SM-Project/E3SM/pull/4752.  That's good to know but not concerning.

xylar commented:
@sbrus89, please don't forget to review this when you can.

sbrus89 commented:
@xylar, will do.

sbrus89 commented:
Now I'm not seeing it. Probably something on my end.

xylar commented on utils/bisect/bisect.py:
```diff
@@ -0,0 +1,92 @@
+def bisect(good, bad, e3sm_path, load_script, config_file, first_parent):
+    """
+    The driver function for calling ``git bisect`` to find the first "bad"
+    commit between a known "good" and "bad" commit.
```
```suggestion
    commit between a known "good" and "bad" E3SM commit.
```

xylar commented on utils/bisect/bisect.py:
```diff
@@ -0,0 +1,92 @@
+
+    Parameters
+    ----------
+    good : str
+        The hash or tag of a "good" commit that passes the test(s) specified
+        in the config file
+    bad : str
+        The hash or tag of a "bad" commit that fails the test(s)
```
```suggestion
    good : str
        The hash or tag of a "good" E3SM commit that passes the test(s) specified
        in the config file
    bad : str
        The hash or tag of a "bad" E3SM commit that fails the test(s)
```

xylar commented on utils/bisect/README.md:
````diff
@@ -0,0 +1,28 @@
+   ```shell
+   cp utils/bisect/example.cfg bisect.cfg
+   ```
+2. Modify the config options with the appropriate "good" and "bad" commit hash
````
```suggestion
2. Modify the config options with the appropriate "good" and "bad" E3SM commit hash
```

xylar commented on utils/bisect/example.cfg:
```diff
@@ -0,0 +1,30 @@
+# config options related to using "git bisect" to find the first commit that
```
```suggestion
# config options related to using "git bisect" to find the first E3SM commit that
```

xylar commented on utils/bisect/example.cfg:
```diff
@@ -0,0 +1,30 @@
+# causes a set of one or more tests to fail
+[bisect]
+
+# The hash or tag of a good commit where the tests pass
+good = 44814ae
+# The hash or tag of a bad commit where the tests fail
+bad = 7b87d1f
```
```suggestion
# The hash or tag of a good E3SM commit where the tests pass
good = 44814ae
# The hash or tag of a bad E3SM commit where the tests fail
bad = 7b87d1f
```

pep8speaks commented:
Hello @xylar! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`utils/bisect/bisect.py`](https://github.com/MPAS-Dev/compass/blob/fb5b94c405ae808bb63cede185c39db917edfc53/utils/bisect/bisect.py):

> [Line 21:80](https://github.com/MPAS-Dev/compass/blob/fb5b94c405ae808bb63cede185c39db917edfc53/utils/bisect/bisect.py#L21): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)

xylar commented on utils/bisect/bisect.py:
```diff
@@ -0,0 +1,92 @@
+    Parameters
+    ----------
+    good : str
+        The hash or tag of a "good" E3SM commit that passes the test(s) specified
+        in the config file
```
```suggestion
        The hash or tag of a "good" E3SM commit that passes the test(s)
        specified in the config file
```

xylar commented:
@sbrus89, any chance of giving this a look in the next few weeks?  It would be nice to have it in.

sbrus89 commented:
@xylar, sure I'll put it on my todo list.

xylar commented:
@sbrus89, I'm not sure what changed but I'm not having any luck with this.  My runs just hang when I try to run on an interactive node and they fail in ways I haven't figured out yet when I try to run a job.

xylar commented:
Oh, I take that back.  The job script worked fine, it just returned an error code for reasons I don't understand.  It seems like it found the bad commit and then tried to run again or something.  I need to investigate sometime but don't have time right now.

sbrus89 commented on utils/bisect/bisect.py:
```diff
@@ -0,0 +1,92 @@
+    else:
+        flags = ''
+
+    commands = f'source {load_script}; ' \
```
Is this first `source {load_script}` necessary?

sbrus89 replied:
I suppose this is to ensure that  the `git` version is recent?

xylar replied:
Yes, I think that was the idea.  Is it causing trouble.

sbrus89 replied:
No, it doesn't cause any trouble. It can just slow things down a bit.

sbrus89 commented on utils/bisect/bisect_step.py:
```diff
@@ -0,0 +1,120 @@
+               f'git submodule update --init --recursive; ' \
+               f'make clean >& {work_path}/clean.log; ' \
+               f'{make_command} >& {work_path}/make.log; ' \
+               f'{setup_command} -p {mpas_path} -w {work_path}; ' \
```
Adding the `--copy_executable` option from #360 would be useful here.

sbrus89 replied:
If you agree, then we should merge #360 before this PR.

xylar replied:
I think it makes the most sense to include this in `setup_command`, which will do in the example.

sbrus89 commented on utils/bisect/bisect_step.py:
```diff
@@ -0,0 +1,120 @@
+    load_script = to_abs(load_script, launch_path)
+
+    commands = f'cd {mpas_path}; ' \
+               f'source {load_script}; ' \
```
Is this `source {load_script}` necessary?

sbrus89 replied:
Same here. Is this to make sure `git` is reasonably up-to-date?

sbrus89 replied:
If so, that's fine. But if we can get rid of it, it would help speed things up a little.

xylar replied:
Okay, we can let it go and assume the user has called `module load git` or something like that if need be.

sbrus89 commented on utils/bisect/bisect_step.py:
```diff
@@ -0,0 +1,120 @@
+               f'{setup_command} -p {mpas_path} -w {work_path}; ' \
+               f'cd {work_path}; ' \
+               f'{run_command}'
+    print('\nRunning:')
```
Before this line, I suggest adding something to the effect of:
```
    print('\n')
    print(72*'-')
    print('Biscect Step')
    print(72*'-')
```
I think this would be helpful in cleaning up the screen output to make it more clear where the bisect steps begin. This isn't quite perfect, because `git bisect run` does the bisect between calls to this script. Maybe those lines should go at the end then. But, then it wouldn't show up for the first call...

xylar replied:
```suggestion
    print('\n')
    print(72*'-')
    print('Biscect Step')
    print(72*'-')
    print('\nRunning:')
```

sbrus89 commented on utils/bisect/bisect_step.py:
```diff
@@ -0,0 +1,120 @@
+    print('\nRunning:')
+    print_commands = commands.replace('; ', '\n  ')
+    print(f'  {print_commands}\n\n')
+    subprocess.check_call(commands, shell=True)
```
Maybe add a `print('\n')` after this line so the `compass run` screen output is separate from the `git bisect run` output

sbrus89 replied:
Although, this only works when a test passes.

xylar replied:
I'll add this after the `run()` call.

sbrus89 commented on utils/bisect/bisect.py:
```diff
@@ -0,0 +1,92 @@
+               f'git bisect bad {bad}; ' \
+               f'git bisect run {cwd}/utils/bisect/bisect_step.py' \
+               f'  -f {config_file}'
+    print('\nRunning:')
```
Before this line maybe add something like:

```
    print('\n')
    print(72*'-')
    print('Biscect Initialization')
    print(72*'-')
```

to clearly indicate (along with the similar comment in `bisect_step.py`) what is going on at the highest level in the process.

xylar replied:
```suggestion
    print('\n')
    print(72*'-')
    print('Biscect Initialization')
    print(72*'-')
    print('\nRunning:')
```

xylar commented on utils/bisect/bisect_step.py:
```diff
@@ -0,0 +1,120 @@
+        work_base=section['work_base'], load_script=section['load_script'],
+        make_command=section['make_command'],
+        setup_command=section['setup_command'],
+        run_command=section['run_command'])
```
```suggestion
        run_command=section['run_command'])
    print ('\n')
```

xylar commented on utils/bisect/bisect.py:
```diff
@@ -0,0 +1,96 @@
+    else:
+        flags = ''
+
+    commands = f'source {load_script}; ' \
+               f'cd {e3sm_path}; ' \
```
```suggestion
    commands = f'cd {e3sm_path}; ' \
```

xylar commented on utils/bisect/bisect_step.py:
```diff
@@ -0,0 +1,125 @@
+    work_base = to_abs(work_base, launch_path)
+    load_script = to_abs(load_script, launch_path)
+
+    commands = f'cd {mpas_path}; ' \
+               f'source {load_script}; ' \
```
```suggestion
    commands = f'cd {mpas_path}; ' \
```

xylar commented:
> The only other comment I have is it may be convenient to create a summary log file that just prints the commit information along with the PASS/FAIL status of each of the commits that is checked-out in the process. I think that is very optional though...

@sbrus89, I don't feel like I have time to do that right now.  I would welcome a follow-up PR to add this, though!

Could you take a look and see if I've addressed your suggestions?  I think so.  If so, I'll merge after CI has finished unless you want to run another test.

xylar commented:
@sbrus89, great, thanks!
